### PR TITLE
[Core] Fix check failure: sync_reactors_.find(reactor->GetRemoteNodeID()) == sync_reactors_.end()

### LINKED ISSUE #45639: [Core] GCS crashed with Check failed: sync_reactors_.find(reactor->GetRemoteNodeID()) == sync_reactors_.end()
### What happened + What you expected to happen

```
2024-05-29 22:51:16	[2024-05-29 22:51:15,987 C 4837 4890] (gcs_server) ray_syncer.cc:248: Check failed: sync_reactors_.find(reactor->GetRemoteNodeID()) == sync_reactors_.end()
```

### Versions / Dependencies

N/A

### Reproduction script

N/A

### Issue Severity

None

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?

The sequence of events that triggers this check failure is

1. There is a transient network error and worker node's `RayClientBidiReactor::OnDone` is called with non-ok grpc status.
2. `cleanup_cb_` of `RayClientBidiReactor` is called which reconnects to the head node.
3. The reconnection causes `RaySyncerService::StartSync` to be called.
4. `RaySyncer::Connect` is called with a new `RayServerBidiReactor`.
5. Because the head node still has the old `RayServerBidiReactor` from the worker node's previous connection, the check fails.

This PR fixes the issue by disconnecting the old `RayServerBidiReactor` first before adding a new one.

## Related issue number
Closes #45639

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_network_failure_e2e.py [hunk 1]
```diff
@@ -124,6 +125,77 @@ def check_task_pending(n=0):
     wait_for_condition(lambda: check_task_pending(2))
 
 
+head2 = gen_head_node(
```
Comment (rkooo567) [inline python/ray/tests/test_network_failure_e2e.py:128]: this test seems very useful. wonder if we should extend it to more e2e tests in some stress tests?

File: src/ray/common/ray_syncer/ray_syncer.cc [hunk 2]
```diff
@@ -350,14 +354,31 @@ ServerBidiReactor *RaySyncerService::StartSync(grpc::CallbackServerContext *cont
...
+        }
         syncer_.sync_reactors_.erase(node_id);
         syncer_.node_state_->RemoveNode(node_id);
       });
   RAY_LOG(INFO).WithField(kLogKeyNodeID, NodeID::FromBinary(reactor->GetRemoteNodeID()))
       << "Get connection";
+  // Disconnect exiting connection if there is any.
+  // This can happen when there is transient network error
+  // and the client reconnects.
+  syncer_.Disconnect(reactor->GetRemoteNodeID());
```
Comment (rkooo567) [inline src/ray/common/ray_syncer/ray_syncer.cc:381]: QQ: why don't we call this in clean_cb?

**Review by rkooo567 — APPROVED** [review]
LG. one minor question

File: python/ray/tests/test_network_failure_e2e.py (hunk 1 above)
Comment (jjyao) [reply python/ray/tests/test_network_failure_e2e.py:128]: Yea, I'll use the same pattern for other tests. I'm also exploring chaos test tools for randomly injecting network errors.

File: src/ray/common/ray_syncer/ray_syncer.cc (hunk 2 above)
Comment (jjyao) [reply src/ray/common/ray_syncer/ray_syncer.cc:381]: `syncer_.Disconnect` triggers the call of `cleanup_cb`

**Pull request merged**